flink/src/main/java/org/apache/iceberg/flink/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.actions;
+
+import java.util.List;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.iceberg.CombinedScanTask;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.actions.BaseRewriteDataFilesAction;
+import org.apache.iceberg.flink.source.RowDataRewriter;
+import org.apache.iceberg.io.FileIO;
+
+public class RewriteDataFilesAction extends BaseRewriteDataFilesAction<RewriteDataFilesAction> {
+
+  private ExecutionEnvironment env;
+
+  public RewriteDataFilesAction(ExecutionEnvironment env, Table table) {
+    super(table);
+    this.env = env;
+  }
+
+  @Override
+  protected FileIO fileIO() {
+    return table().io();
+  }
+
+  @Override
+  protected List<DataFile> rewriteDataForTasks(List<CombinedScanTask> combinedScanTasks) {
+    DataSet<CombinedScanTask> dataSet = env.fromCollection(combinedScanTasks).setParallelism(combinedScanTasks.size());
```

Comment:
I think `setParallelism` may not works.
What we want is a `SparkContext.parallelize`, but this `setParallelism` just set parallelism, records still in a single node, downstream operators only have computation on one node.
Can you check this?

Reply:
> Thanks @zhangjun0x01 for your contribution, why do you want to use `DataSet` API? `DataSet` API is being deprecated.
> Looks like just need a `parallelize` and `map` and `collect`. I think `DataStream`(We can call it BoundedStream) can finish this.
> What do you think?

Hi,@JingsongLi：
I look up the api of the DataStream, it seems  no `collect` method

Reply:
You can use `DataStreamUtils.collect`

Reply:
I updated the pr, use DataStream instead of DataSet , and set the parallelism to StreamExecutionEnvironment.

but setParallelism for  `env.fromCollection(combinedScanTasks)`  does not seem to have an impact on downstream operator,this is my test job graph for use dataset.   https://blog.csdn.net/zhangjun5965/article/details/109469798

Reply:
Thanks for the update and check, I think you are right, the graph is good.

Reply:
I think about parallelism again.
Can we not set parallelism? Flink is a push based model, which is not suitable for taking on too much parallelism.
So can we just let it as default parallelims?

Reply:
If we do not set the parallelism, we can use the default parallelism of flink. If the user wants to modify the parallelism to improve the execution speed, it can be set by the `-p` parameter. However, the user may not know how much parallelism to set，because users may not know how many data files to Rewrite each time. If the setting is too large, it may cause a waste of resources. 

So I think we can  set a max parallelism. If the size of CombinedScanTask list  is less than max parallelism, we use the CombinedScanTask list size as the job parallelism, if the size  is greater than max parallelism，we use the max parallelism which we set.
What do you think?

Reply:
Let me explain why we can't set parallelism directly:
Flink is in pipeline mode by default, which means that all tasks must be started.
If we want to compact a large table, it may require thousands of parallelism, which leads to the need for a very large cluster.

This is similar to the Hive source parallelism inference: https://ci.apache.org/projects/flink/flink-docs-master/dev/table/hive/hive_read_write.html#source-parallelism-inference
The value should be used to infer parallelism, but we need a max parallelism.

So I agree with you very much. Let's add a max parallelism.

Reply:
BTW, you should set parallelism after `map` instead of modifying env.

Reply:
I updated the pr, although we set the parallelism for StreamExecutionEnvironment, I found that the DataStreamUtils.collect operator only uses 1 parallelism, so set parallelism to map operator or StreamExecutionEnvironment seems to have the same effect, but set parallelism to map is better to understand .

I submit an issue (#1660) to  set max parallelism a few days ago, I will open a new pr  for #1660 when this PR is merged

Reply:
> although we set the parallelism for StreamExecutionEnvironment, I found that the DataStreamUtils.collect operator only uses 1 parallelism, so set parallelism to map operator or StreamExecutionEnvironment seems to have the same effect, but set parallelism to map is better to understand.

It is better not to bring extra impact to env, because this env may be reused by users.

> I submit an issue (#1660) to set max parallelism a few days ago, I will open a new pr for #1660 when this PR is merged

Can you modify in this PR? Or just not set parallelism in this PR?
I think it is not good to bring user cluster risk to master code.

Reply:
I  update the code , add the max parallelism in this PR.